add E2E tests for DAG Runs tab

Sahil-Shadwal opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

Add E2E tests for DAG Runs tab functionality.

## Changes
- Added `data-testid` to [DagRuns.tsx](cci:7://file:///Users/sahilshadwal/Desktop/airflow/airflow-core/src/airflow/ui/src/pages/DagRuns.tsx:0:0-0:0) for E2E testing
- Extended [DagsPage.ts](cci:7://file:///Users/sahilshadwal/Desktop/airflow/airflow-core/src/airflow/ui/tests/e2e/pages/DagsPage.ts:0:0-0:0) with 7 methods for Runs tab interactions
- Created test suite with 4 test cases covering navigation, display, filtering, search, and pagination

## Test Coverage
✅ Navigate to Runs tab  
✅ Verify run details (ID, state, times, duration)  
✅ Click run to view details  
✅ Filter and search functionality  
✅ Pagination

Closes #59541 

Related issues: #59028

<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

Sahil-Shadwal committed:
add E2E tests for DAG Runs tab

vatsrahul1001 committed:
Merge branch 'main' into feat--e2e-dag-06

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/pages/DagsPage.ts:
```diff
@@ -116,6 +121,115 @@ export class DagsPage extends BasePage {
   /**
    * Navigate to details tab and verify Dag details are displayed correctly
```
The first comment is orphaned from a previous method. We can remove it.

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +84,87 @@ test.describe("Dag Details Tab", () => {
+
+    const runs = await dagsPage.getRunDetails();
+
+    expect(runs.length).toBeGreaterThan(0);
```
If the test DAG has no runs, all tests fail. Consider Triggering a DAG run before testing

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/pages/DagsPage.ts:
```diff
@@ -116,6 +121,115 @@ export class DagsPage extends BasePage {
+    await searchInput.fill(searchTerm);
+
+    // Wait for the search to take effect
+    await this.page.waitForTimeout(1000);
```
Wait for a specific condition instead of arbitrary timeout.

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +84,87 @@ test.describe("Dag Details Tab", () => {
+/*
+ * DAG Runs Tab E2E Tests
+ */
+test.describe("DAG Runs Tab", () => {
```
Use Dag instead of DAG

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/pages/DagsPage.ts:
```diff
@@ -116,6 +121,115 @@ export class DagsPage extends BasePage {
+  /**
+   * Filter runs by state
+   */
+  public async filterByState(state: string): Promise<void> {
```
We are not using this method in any test

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/pages/DagsPage.ts:
```diff
@@ -116,6 +121,115 @@ export class DagsPage extends BasePage {
+  /**
+   * Search for dag runs by run ID pattern
+   */
+  public async searchRun(searchTerm: string): Promise<void> {
```
We are not using searchRun metod in any test

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +84,87 @@ test.describe("Dag Details Tab", () => {
+    // Check if next button is enabled (has more than one page)
+    const isNextEnabled = await dagsPage.paginationNextButton.isEnabled();
+
+    if (isNextEnabled) {
```
If there's only one page of results, the core pagination test is skipped silently. Consider using test.skip() with a message.

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/pages/DagsPage.ts:
```diff
@@ -116,6 +121,115 @@ export class DagsPage extends BasePage {
+      const cells = row.locator("td");
+
+      // Assuming first column is run ID and state column exists
+      const runId = (await cells.nth(0).textContent()) ?? "";
```
We could use better selector here
Assumes runId is column 0 and state is column 1. If UI changes column order, test breaks.

Sahil-Shadwal committed:
fix reviewed changes and test failures

vatsrahul1001 committed:
Merge branch 'main' into feat--e2e-dag-06

vatsrahul1001 committed:
Merge branch 'main' into feat--e2e-dag-06

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +84,142 @@ test.describe("Dag Details Tab", () => {
+
+  const testDagId = testConfig.testDag.id;
+
+  test.beforeEach(async ({ page }) => {
```
Let's replace it with beforeALL as mentioned in https://github.com/apache/airflow/pull/59754#issuecomment-3713727824

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +84,142 @@ test.describe("Dag Details Tab", () => {
     await dagsPage.verifyDagDetails(testDagId);
   });
 });
```
Do not add these test to dags-list.spec.ts add these to dag-runs.spec.ts

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +84,142 @@ test.describe("Dag Details Tab", () => {
+    expect(firstSearchedRun.runId).toContain(targetRunId);
+  });
+
+  test("should verify pagination works on the Runs tab", async () => {
```
rename test to `verify pagination`

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +84,142 @@ test.describe("Dag Details Tab", () => {
+    await dagsPage.triggerDag(testDagId);
+  });
+
+  test("should navigate to Runs tab and display correctly", async () => {
```
Rename test to `verify runs table displays with valid data`

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +84,142 @@ test.describe("Dag Details Tab", () => {
+    expect(runs.length).toBeGreaterThan(0);
+  });
+
+  test("should verify run details are displayed correctly", async () => {
```
This test looks redundant to me

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +84,142 @@ test.describe("Dag Details Tab", () => {
+    expect(firstRun.state).toBeTruthy();
+  });
+
+  test("should click run and navigate to details page", async () => {
```
Rename test to `verify run details page navigation`

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +84,142 @@ test.describe("Dag Details Tab", () => {
+    await dagsPage.verifyRunDetailsPage(firstRun.runId);
+  });
+
+  test("should filter runs by state", async () => {
```
Rename test to `verify filtering`. Also, we should filter with different type like state, run_id, run_type.

vatsrahul1001 replied:
Also ensure to validate filtering works and count changes

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +84,142 @@ test.describe("Dag Details Tab", () => {
+    expect(filteredRuns.every((run) => run.state === targetState)).toBeTruthy();
+  });
+
+  test("should search runs", async () => {
```
I think we do not need this test, this can be covered in filtering test. when we apply filter with run_id it does same thing

vatsrahul1001 commented on airflow-core/src/airflow/ui/tests/e2e/specs/dags-list.spec.ts:
```diff
@@ -84,3 +84,142 @@ test.describe("Dag Details Tab", () => {
+    // Check if next button is enabled (has more than one page)
+    const isNextEnabled = await dagsPage.paginationNextButton.isEnabled();
+
+    if (!isNextEnabled) {
```
No need to skip, you can use approach which I have mentioned here https://github.com/apache/airflow/pull/59754#issuecomment-3713690469

Sahil-Shadwal committed:
addressed alll feedback and passed all tests

Sahil-Shadwal committed:
Merge branch 'main' into feat--e2e-dag-06

Sahil-Shadwal committed:
feat: Add comprehensive E2E tests for DAG runs page

- Add tests for run details navigation and verification
- Implement filtering tests for run state and type
- Add search functionality tests
- Implement pagination tests with offset/limit
- Add sorting tests for logical date column
- Create reusable DagsPage helper methods for DAG run operations

Sahil-Shadwal committed:
fix ci

Sahil-Shadwal committed:
fix: Remove LoginPage usage from dag-runs.spec.ts to match successful PR pattern

- Remove LoginPage import and all login-related code
- Tests now receive authenticated page from Playwright's global storageState
- Initialize DagsPage directly in each test function
- Matches pattern from successfully merged PRs: #59943, #59919, #59734
- Fixes CI failures across Chromium, Firefox, and WebKit

Addresses reviewer feedback from @vatsrahul1001

Sahil-Shadwal committed:
Merge branch 'main' into feat--e2e-dag-06

Resolved conflicts by keeping our DAG-specific runs tab test (DAG-006).
Main branch has a different dag-runs.spec.ts for global DAG Runs page (DAGRUNS-001).

vatsrahul1001 committed:
Merge branch 'main' into feat--e2e-dag-06
